### PR TITLE
Add SetBatchSize pass to MLIR converter

### PR DESCRIPTION
**Pull request by Tombana**
## What do these changes do?
This adds a pass to the converter that sets the dynamic batchsize to 1 on the input layer.

When `batch_size` is not explicitly set to 1, it will remain a `?` (wildcard) throughout all MLIR passes. Only at the final MLIR to Flatbuffer conversion stage, are all the `?` simply converted to `1`:
https://github.com/tensorflow/tensorflow/blob/v2.7.0/tensorflow/compiler/mlir/lite/flatbuffer_export.cc#L844

This causes broken Reshape and Concatenation ops: the MLIR passes will correctly infer the Reshape output shape to something like `?x?x2`, and then the flatbuffer exporter simply overrides both with `1`. Other artifacts
- left-over `StridedSlice` ops in the graph (which should've been folded out)
- invalid Reshape shapes: image
- invalid concatenation shapes: image 
- I forgot the exact details but something like depthwise layers with dilation also got shape mismatches.

The graphdef converter code has a `batch_size=1` fix in Python, but the saved model converter did not have something like this yet.

## How Has This Been Tested?
MLIR filecheck tests have been added.

### CODE REVIEW

File: larq_compute_engine/mlir/tests/batchsize.mlir [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  // CHECK: %arg0: tensor<1x6xf32>
+  // CHECK: %arg1: tensor<1x4xf32>
+}
+
+// This is the same model, but one of the two inputs has been given a fixed batch size in Python
+
+// CHECK-LABEL: @dual_input_one_fixed_size
+func @dual_input_one_fixed_size(%arg0: tensor<?x6xf32> {tf_saved_model.index_path = ["input_2"]}, %arg1: tensor<1x4xf32> {tf_saved_model.index_path = ["input_1"]}, %arg2: tensor<!tf.resource<tensor<6xf32>>> {tf_saved_model.bound_input = @"dense/bias"}, %arg3: tensor<!tf.resource<tensor<4x6xf32>>> {tf_saved_model.bound_input = @"dense/kernel"}) -> (tensor<?x6xf32> {tf_saved_model.index_path = ["tf.__operators__.add"]}) attributes {tf.entry_function = {control_outputs = "", inputs = "serving_default_input_2:0,serving_default_input_1:0", outputs = "StatefulPartitionedCall:0"}, tf_saved_model.exported_names = ["serving_default"]} {
+  %0 = "tf.ReadVariableOp"(%arg2) {device = ""} : (tensor<!tf.resource<tensor<6xf32>>>) -> tensor<6xf32>
+  %1 = "tf.ReadVariableOp"(%arg3) {device = ""} : (tensor<!tf.resource<tensor<4x6xf32>>>) -> tensor<4x6xf32>
```
Comment (lgeiger) [inline larq_compute_engine/mlir/tests/batchsize.mlir:29]: Let's remove `tf.ReadVariableOp` and `tf.Identity` ops to improve readability of the test case.

File: larq_compute_engine/mlir/transforms/batchsize.cc [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+}  // namespace
+
+// Creates an instance of the ZeroPointCompatibility pass.
+std::unique_ptr<OperationPass<FuncOp>> CreateSetBatchSizePass() {
+  return std::make_unique<SetBatchSizePass>();
+}
+
+static PassRegistration<SetBatchSizePass> pass("mlir-setbatchsize",
+                                               "Set batch size to 0");
```
Comment (lgeiger) [inline larq_compute_engine/mlir/transforms/batchsize.cc:66]:
```suggestion
                                               "Set batch size to 1");
```

**Review by lgeiger — COMMENTED** [review]
Very cool, this is really useful!

I just have a few minor suggestions to improve readability, other than that this looks great!

File: larq_compute_engine/mlir/tf_tfl_passes.cc [hunk 3]
```diff
@@ -92,6 +92,10 @@ void AddTFToLCETFLConversionPasses(
   pass_manager->addPass(
       mlir::tf_saved_model::CreateOptimizeGlobalTensorsPass());
 
+  // Set the batch size of the function input to 1 and let shape inference
+  // propagate this in the next pass.
+  pass_manager->addPass(mlir::CreateSetBatchSizePass());
```
[the remark this answers is not in the document]
Comment (Tombana) [reply larq_compute_engine/mlir/tf_tfl_passes.cc:97]: Originally I just applied it outside of the pass manager, directly on the `module` object that we get from `ImportSavedModel`. Then I decided a pass was cleaner because it also allows simple FileCheck tests.

First I placed the pass right before the first `tf-shape-inference` pass, but at that point there were several `@func`'s in the IR with these `StateFullPartitionedCall`s in them, so it was not clear whether this fix should be applied to all functions or not.
Then I placed it right before the second `tf-shape-inference` pass, at which point the `StateFullPartitionedCall`s had dissapeared and things were in the TF dialect.

So in conclusion: it can probably be placed in many locations, but the earlier the better since then more shapes will be known and more optimizations can be applied.

File: larq_compute_engine/mlir/tests/batchsize.mlir [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  %6 = "tf.Identity"(%5) {device = ""} : (tensor<?x6xf32>) -> tensor<?x6xf32>
+  return %6 : tensor<?x6xf32>
+  // CHECK: %arg0: tensor<1x6xf32>
+  // CHECK: %arg1: tensor<1x4xf32>
+}
+
+// This is the same model, but one of the two inputs has been given a fixed batch size in Python
+
+// CHECK-LABEL: @dual_input_one_fixed_size
+func @dual_input_one_fixed_size(%arg0: tensor<?x6xf32> {tf_saved_model.index_path = ["input_2"]}, %arg1: tensor<1x4xf32> {tf_saved_model.index_path = ["input_1"]}, %arg2: tensor<!tf.resource<tensor<6xf32>>> {tf_saved_model.bound_input = @"dense/bias"}, %arg3: tensor<!tf.resource<tensor<4x6xf32>>> {tf_saved_model.bound_input = @"dense/kernel"}) -> (tensor<?x6xf32> {tf_saved_model.index_path = ["tf.__operators__.add"]}) attributes {tf.entry_function = {control_outputs = "", inputs = "serving_default_input_2:0,serving_default_input_1:0", outputs = "StatefulPartitionedCall:0"}, tf_saved_model.exported_names = ["serving_default"]} {
```
[the remark this answers is not in the document]
Comment (Tombana) [reply larq_compute_engine/mlir/tests/batchsize.mlir:27]: That is true, but since the new pass is overriding these `mlir::Value` objects, I wanted to make sure that this extra metadata is not accidentally destroyed and that the pass also works on "real" IR coming from Python models. That's why I chose a dump from a real model instead of a simpler test.

File: larq_compute_engine/mlir/transforms/batchsize.cc (hunk 2 above)
Comment (Tombana) [reply larq_compute_engine/mlir/transforms/batchsize.cc:66]: Oops :sweat_smile:

File: larq_compute_engine/mlir/tf_tfl_passes.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq_compute_engine/mlir/tf_tfl_passes.cc:97]: Sounds good! Thanks for the explanation!

File: larq_compute_engine/mlir/tests/batchsize.mlir (hunk 4 above)
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq_compute_engine/mlir/tests/batchsize.mlir:27]: Makes sense 👍

[the remark this answers is not in the document]
Comment (Tombana) [reply larq_compute_engine/mlir/tests/batchsize.mlir]: I've now added a simpler test at the top without all the extra parts. I've left this test in here, but with an extra `CHECK` to see if all the attributes are unchanged.

**Review by lgeiger — APPROVED** [review]
